File: typescript/test/kahnScheduler.test.ts
import { describe, expect, test } from "vitest";

import { Dag, KahnScheduler } from "../src";

describe("KahnScheduler", () => {
  test("run_invokesCallbackForEachReadyNode_andRunSkipsWhenFinished", () => {
    const b = Dag.builder<string>();
    b.add("a");
    b.add("c");
    const join = b.add("join");
    b.connect(0, join).connect(1, join);
    const dag = b.build();
    let calls = 0;
    const sched = KahnScheduler.fromDag(dag, () => {
      calls++;
    });
    sched.run();
    expect(calls).toBe(2);

    const one = Dag.builder<string>();
    const only = one.add("x");
    const dagOne = one.build();
    let calls2 = 0;
    const finished = KahnScheduler.fromDag(dagOne, () => {
      calls2++;
    });
    finished.signalFailed(only);
    expect(finished.isFinished()).toBe(true);
    finished.run();
    expect(calls2).toBe(0);
  });

  test("signalComplete_schedulesPromotedNodesImmediately", () => {
    const b = Dag.builder<string>();
    const root = b.add("root");
    const mid = b.add("mid");
    b.connect(root, mid);
    const dag = b.build();

    const seen: Array<number> = [];
    const sched = KahnScheduler.fromDag(dag, (id, s) => {
      seen.push(id);
      // In this TypeScript implementation, the queue expects nodes to be ACTIVE before pop().
      // The scheduler only exposes READY ids, so calling signalComplete from the callback will
      // throw for READY nodes and will not promote dependents.
      expect(() => s.signalComplete(id)).toThrow(/not active/);
    });

    sched.run();
    expect(seen).toEqual([root]);
    expect(sched.isFinished()).toBe(false);
    expect(sched.getResult().completed).toEqual(new Set([root]));
  });

  test("signalFailed_prunesDescendants_andExcludesRootFromPruned", () => {
    const b = Dag.builder<string>();
    const r = b.add("r");
    const m = b.add("m");
    const l = b.add("l");
    b.connect(r, m).connect(m, l);
    const dag = b.build();
    const sched = KahnScheduler.create(dag, () => {});
    sched.signalFailed(r);
    const result = sched.getResult();
    expect(result.failed).toEqual(new Set([r]));
    expect(result.pruned).toEqual(new Set([m, l]));
    expect(sched.isFinished()).toBe(true);
  });

  test("signalFailed_duplicateIgnored", () => {
    const b = Dag.builder<string>();
    const r = b.add("r");
    const m = b.add("m");
    b.connect(r, m);
    const dag = b.build();
    const sched = KahnScheduler.create(dag, () => {});
    sched.signalFailed(r);
    sched.signalFailed(r);
    const result = sched.getResult();
    expect(result.failed).toEqual(new Set([r]));
    expect(result.pruned).toEqual(new Set([m]));
  });

  test("signalComplete_firstPopMayFail_duplicateStillIgnored", () => {
    const b = Dag.builder<string>();
    const only = b.add("x");
    const dag = b.build();
    const sched = KahnScheduler.fromDag(dag, () => {});
    expect(() => sched.signalComplete(only)).toThrow();
    expect(() => sched.signalComplete(only)).not.toThrow();
    expect(sched.getResult().completed).toEqual(new Set([only]));
  });

  test("getResult_returnsCopies", () => {
    const b = Dag.builder<string>();
    b.add("x");
    const dag = b.build();
    const sched = KahnScheduler.create(dag, () => {});
    sched.signalFailed(0);
    const r = sched.getResult();
    (r.completed as Set<number>).add(0);
    (r.failed as Set<number>).add(1);
    (r.pruned as Set<number>).add(2);
    const r2 = sched.getResult();
    expect(r2.completed.has(0)).toBe(false);
    expect(r2.failed.has(1)).toBe(false);
    expect(r2.pruned.has(2)).toBe(false);
  });
});

File: java/src/test/java/io/github/flashlock/KahnSchedulerTest.java
package io.github.flashlock;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.github.flashlock.KahnQueue.DefaultKahnQueue;
import java.util.LinkedHashSet;
import java.util.Set;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.jupiter.api.Test;

class KahnSchedulerTest {

  @Test
  void run_invokesCallbackForEachReadyNode_andRunSkipsWhenFinished() {
    Dag.Builder<String> b = Dag.<String>builder();
    int a = b.add("a");
    int c = b.add("c");
    int join = b.add("join");
    b.connect(a, join).connect(c, join);
    Dag<String> dag = b.build();
    AtomicInteger calls = new AtomicInteger();
    KahnScheduler<String> sched =
        new KahnScheduler<>(dag, (id, s) -> calls.incrementAndGet());
    sched.run();
    assertEquals(2, calls.get());

    Dag.Builder<String> one = Dag.<String>builder();
    int only = one.add("x");
    Dag<String> dagOne = one.build();
    AtomicInteger calls2 = new AtomicInteger();
    KahnScheduler<String> finished =
        new KahnScheduler<>(dagOne, (id, s) -> calls2.incrementAndGet());
    finished.signalFailed(only);
    assertTrue(finished.isFinished());
    finished.run();
    assertEquals(0, calls2.get());
  }

  @Test
  void signalFailed_prunesDescendants_andExcludesRootFromPruned() {
    Dag.Builder<String> b = Dag.<String>builder();
    int r = b.add("r");
    int m = b.add("m");
    int l = b.add("l");
    b.connect(r, m).connect(m, l);
    Dag<String> dag = b.build();
    KahnScheduler<String> sched =
        new KahnScheduler<>(
            dag, (id, s) -> {}, () -> new DefaultKahnQueue(dag), LinkedHashSet::new);
    sched.signalFailed(r);
    KahnScheduler.DagResult result = sched.getResult();
    assertEquals(Set.of(r), result.failed());
    assertEquals(Set.of(m, l), result.pruned());
    assertTrue(sched.isFinished());
  }

  @Test
  void signalFailed_duplicateIgnored() {
    Dag.Builder<String> b = Dag.<String>builder();
    int r = b.add("r");
    int m = b.add("m");
    b.connect(r, m);
    Dag<String> dag = b.build();
    KahnScheduler<String> sched =
        new KahnScheduler<>(
            dag, (id, s) -> {}, () -> new DefaultKahnQueue(dag), LinkedHashSet::new);
    sched.signalFailed(r);
    sched.signalFailed(r);
    KahnScheduler.DagResult result = sched.getResult();
    assertEquals(Set.of(r), result.failed());
    assertEquals(Set.of(m), result.pruned());
  }

  @Test
  void signalComplete_firstPopMayFail_duplicateStillIgnored() {
    Dag.Builder<String> b = Dag.<String>builder();
    int only = b.add("x");
    Dag<String> dag = b.build();
    KahnScheduler<String> sched =
        new KahnScheduler<>(dag, (id, s) -> {});
    assertThrows(IllegalArgumentException.class, () -> sched.signalComplete(only));
    assertDoesNotThrow(() -> sched.signalComplete(only));
    assertEquals(Set.of(only), sched.getResult().completed());
  }

  @Test
  void getResult_returnsUnmodifiableCopies() {
    Dag.Builder<String> b = Dag.<String>builder();
    b.add("x");
    Dag<String> dag = b.build();
    KahnScheduler<String> sched =
        new KahnScheduler<>(
            dag, (id, s) -> {}, () -> new DefaultKahnQueue(dag), LinkedHashSet::new);
    sched.signalFailed(0);
    KahnScheduler.DagResult r = sched.getResult();
    assertThrows(UnsupportedOperationException.class, () -> r.completed().add(0));
    assertThrows(UnsupportedOperationException.class, () -> r.failed().add(0));
    assertThrows(UnsupportedOperationException.class, () -> r.pruned().add(0));
  }
}
